DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/src/dsp_control.cpp

```diff
@@ -0,0 +1,148 @@
+// e.g., $ONEAPI_ROOT/dev-utilities/include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+using namespace sycl;
+
+float subtract(float a, float b) { return a - b; }
+
+// Forward declare the kernel names in the global scope.
+// This FPGA best practice reduces name mangling in the optimization reports.
+class GlobalControl;
+class LocalControlPropagateOn;
+class LocalControlPropagateOff;
+
+// Runs the Kernel.
+void KernelRun(const std::vector<float> &input_data,
+               std::vector<float> &output_data_add,
+               std::vector<float> &output_data_sub) {
+
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector device_selector;
+#else
+  ext::intel::fpga_selector device_selector;
+#endif
+
+  try {
+    // Create the SYCL device queue.
+    queue q(device_selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer input_buffer(input_data);
+    buffer output_add_buffer(output_data_add);
+    buffer output_sub_buffer(output_data_sub);
+
+    q.submit([&](handler &h) {
+      accessor input_a(input_buffer, h, read_only);
+      accessor output_add_a(output_add_buffer, h, write_only, no_init);
+      accessor output_sub_a(output_sub_buffer, h, write_only, no_init);
+
+      // Kernel that demonstrates DSP global control.
+      h.single_task<GlobalControl>([=]() [[intel::kernel_args_restrict]] {
+        // Command-line option `-Xsdsp-mode=prefer-softlogic` controls both
+        // addition and subtraction to be implemented in soft-logic.
+        output_add_a[0] = input_a[0] + input_a[1];
+        output_sub_a[0] = subtract(input_a[0], input_a[1]);
+      });
+    });
+
+    q.submit([&](handler &h) {
+      accessor input_a(input_buffer, h, read_only);
+      accessor output_add_a(output_add_buffer, h, write_only, no_init);
+      accessor output_sub_a(output_sub_buffer, h, write_only, no_init);
+
+      // Kernel that demonstrates DSP local control with Propagate::On.
+      h.single_task<LocalControlPropagateOn>([=
+      ]() [[intel::kernel_args_restrict]] {
+        // The local control library function overrides the global control.
+        // Because the Propagate argument is On, not only the addition directly
+        // in the lambda, but also the subtraction in the subtract() function
+        // call inside the lambda are affected by the local control and will be
+        // implemented in DSP.
+        ext::intel::math_dsp_control<>([&] {
+          output_add_a[1] = input_a[0] + input_a[1];
+          output_sub_a[1] = subtract(input_a[0], input_a[1]);
+        });
+      });
+    });
+
+    q.submit([&](handler &h) {
+      accessor input_a(input_buffer, h, read_only);
+      accessor output_add_a(output_add_buffer, h, write_only, no_init);
+      accessor output_sub_a(output_sub_buffer, h, write_only, no_init);
+
+      // Kernel that demonstrates DSP local control with Propagate::Off.
+      h.single_task<LocalControlPropagateOff>([=
```

Comment:
It would be nice to find a config option to clang-format which would change the line-break rules here.   Visually it would be better to put the line-break after [=]() rather than after the =.

Reply:
Both Google style and LLVM style gave me this line-break. Maybe we can live with it?

Reply:
I tried some time ago to find a work around that doesn't involve manually editing after clang-format, and couldn't come up with anything better than this.  This is the formatting used in all our other samples, I agree it's a bit ugly, but I think we have to continue to live with it.